### PR TITLE
feat(ui): Keep stable references to lazy components

### PR DESCRIPTION
**Pull request by scttcper**
This pr pulls the `lazy()` call out of the render path.

In concurrent mode, a spinner is sometimes shown on route change even though the component has already been loaded. We should try to keep stable references to `lazy(() => import('./whatever'));` to benefit from the caching react lazy does and avoid the tick from .then() that is needed to resolve a fresh promise, even if webpack has cached the result already.

[From the docs](https://react.dev/reference/react/lazy#lazy)
> Call lazy outside your components to declare a lazy-loaded React component:

example: go to the subscription page https://sentry.sentry.io/settings/billing/overview/ and flick between the tabs

before - spinner every time you switch tabs even after the component is loaded

https://github.com/getsentry/sentry/assets/1400464/c955c438-4f8f-4caa-93f5-ee1eb795edf7

after - spinner or flicker only when the bundle is loaded the first time. this could be prefetched

https://github.com/getsentry/sentry/assets/1400464/4a75cca4-f353-4008-b9ca-b4b2d4e21ca3

part of https://github.com/getsentry/frontend-tsc/issues/63

### CODE REVIEW

File: static/app/routes.tsx [hunk 1]
```diff
@@ -39,11 +40,12 @@ const SafeLazyLoad = errorHandler(LazyLoad);
 export function makeLazyloadComponent<C extends React.ComponentType<any>>(
   resolve: () => Promise<{default: C}>
 ) {
+  const LazyComponent = lazy<C>(() => retryableImport(resolve));
```
Comment (scttcper) [inline static/app/routes.tsx:43]: pulling this outside the RouteLazyLoad ensures that as long as makeLazyloadComponent isn't called multiple times this should be stable

**Review by evanpurkhiser — APPROVED** [review]

File: static/app/components/lazyLoad.tsx [hunk 2]
```diff
@@ -31,17 +37,24 @@ type Props<C extends ComponentType> = React.ComponentProps<C> & {
...
+ * <LazyLoad LazyComponent={LazyComponent} someComponentProps={...} />
  */
 function LazyLoad<C extends ComponentType>({
   component,
   loadingFallback,
+  LazyComponent,
   ...props
 }: Props<C>) {
-  const LazyComponent = useMemo(
-    () => lazy<C>(() => retryableImport(component)),
```
Comment (scttcper) [inline static/app/components/lazyLoad.tsx:42]: this is what i'm trying to avoid, a new lazy() every time the component is rendered

**Review by ryan953 — APPROVED** [review]

File: static/app/components/lazyLoad.tsx [hunk 3]
```diff
@@ -31,17 +37,24 @@ type Props<C extends ComponentType> = React.ComponentProps<C> & {
  * LazyLoad is used to dynamically load codesplit components via a `import`
  * call. This is primarily used in our routing tree.
  *
- * <LazyLoad component={() => import('./myComponent')} someComponentProps={...} />
+ * Outside the render path
+ * Const LazyComponent = lazy(() => import('./myComponent'))
```
Comment (ryan953) [inline static/app/components/lazyLoad.tsx:41]:
```suggestion
 * const LazyComponent = lazy(() => import('./myComponent'))
```

**Review by ryan953 — COMMENTED** [review]
`const LazyComponent = lazy(() => import('./myComponent'))` makes the most sense.

**Pull request merged**